src/main/java/org/jabref/logic/importer/fetcher/ISIDOREFetcher.java

```diff
@@ -0,0 +1,249 @@
+        if (list.getLength() == 1) {
+            return list.item(0).getTextContent();
+        }
+
+        // If there are multiple abstracts, return the one with the english tag
+        for (int i = 0; i < list.getLength(); i++) {
+            Element abstractElement = (Element) list.item(i);
+            String langAttribute = abstractElement.getAttribute("xml:lang");
+            if (Objects.equals(langAttribute, "en")) {
+                return abstractElement.getTextContent();
+            }
+        }
+        return "";
+    }
+
+    // Get the type of the document, ISIDORE only seems to have select types, also their types are different to
+    // those used by Jabaref.
+    private EntryType getType(NodeList list) {
+        for (int i = 0; i < list.getLength(); i++) {
+            String type = list.item(i).getTextContent();
+            if (type.contains("article") || type.contains("Article")) {
+                return StandardEntryType.Article;
+            }
+            if (type.contains("thesis") || type.contains("Thesis")) {
+                return StandardEntryType.Thesis;
+            }
+            if (type.contains("book") || type.contains("Book")) {
+                return StandardEntryType.Book;
+            }
+        }
+        return StandardEntryType.Misc;
+    }
+
+    // Gets all the authors, separated with the word "and"
+    // For some reason the author field sometimes has extra numbers and letters.
+    private String getAuthor(Node itemElement) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int i = 1; i < itemElement.getChildNodes().getLength(); i += 2) {
+            String next = removeMultipleSpaces(removeNumbers(itemElement.getChildNodes().item(i).getTextContent()));
+            next = next.replace("\n", "");
+            if (next.isBlank()) {
+                continue;
+            }
+            stringBuilder.append(next);
+            stringBuilder.append(" and ");
+        }
+        return removeMultipleSpaces(stringBuilder.substring(0, stringBuilder.length() - 5)).trim();
+    }
+
+    private String removeMultipleSpaces(String in) {
+        int previous = 0;
+        while (spaces(in) != previous) {
+            previous = spaces(in);
+            in = in.replace("  ", " ");
+        }
+        return in;
+    }
+
+    // count the number of spaces in a string
+    private int spaces(String s) {
+        int x = 0;
+        for (int i = 0; i < s.length(); i++) {
+            if (s.charAt(i) == ' ') {
+                x++;
+            }
+        }
+        return x;
+    }
+
+    // Remove numbers from a string and everything after the number, (helps with the author field).
+    private String removeNumbers(String string) {
+        for (int i = 0; i < string.length(); i++) {
+            if (Character.isDigit(string.charAt(i))) {
+                return string.substring(0, i);
+            }
+        }
+        return string;
```

Comment:
```suggestion
        return string.replaceFirst("\\d.*", "");
```

Reply:
Is this suggestion OK for you or do you see any issues?

Reply:
Unfortunately some of the Author nodes are a bit weird. Sometimes they contain a string of numbers and a dash after the name and then repeat the name again for no apparent reason e.g. (Patrick Bonnel becomes Patrick Bonnel 0766-05442 Patrick). So to solve this I simply removed everything after the first number.

Reply:
I have a second ChatGPT suggestion, but I don't know about performance gains. I tend to keep the above suggestion and optimize if there are performance issues

```java
import java.util.regex.*;

private String removeNumbers(String string) {
    Matcher m = Pattern.compile("^[^\\d]*").matcher(string);
    if (m.find()) {
        return m.group(0);
    }
    return string;
}
```

---

Given your context, the method you've provided returns the portion of the string before the first number. Using a regular expression, we can accomplish the same task more concisely.

Here's a refactored version of the `removeNumbers` method using regex:

```java
import java.util.regex.*;

private String removeNumbers(String string) {
    Matcher m = Pattern.compile("^[^\\d]*").matcher(string);
    if (m.find()) {
        return m.group(0);
    }
    return string;
}
```

The regular expression `^[^\\d]*` can be interpreted as:

- `^` asserts position at the start of a string.
- `[^\\d]*` matches zero or more non-digit characters.

The method works by matching as many non-digit characters as possible from the beginning of the string until it encounters a digit (or the end of the string). If a match is found, it returns that match; otherwise, it simply returns the original string.

Reply:
The second option is good, if `Pattern.compile(....)` is moved to a class constant.